realm/realm-library/src/main/java/io/realm/RealmCache.java

```diff
@@ -117,6 +119,17 @@ private RealmCache(RealmConfiguration config) {
 
         RefAndCount refAndCount = cache.refAndCountMap.get(RealmCacheType.valueOf(realmClass));
 
+        if (refAndCount.globalCount == 0) {
+            SharedRealm sharedRealm = SharedRealm.getInstance(configuration);
+            sharedRealm.beginTransaction();
+            if (Table.migratePrimaryKeyTableIfNeeded(sharedRealm)) {
```

Comment:
Is it possible to make a check if we need to migrate outside a transaction. Transactions on init are pretty dangerous in multi-process scenarios, since they most likely happen on the UI thread.